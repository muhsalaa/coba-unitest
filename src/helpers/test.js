export function hexToRGBString(hex) {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? `rgb(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
        result[3],
        16
      )})`
    : null;
}

export const findByTestAttr = (wrapper, val, elem = '') => {
  return wrapper.find(`${elem}[data-testid='${val}']`);
};
