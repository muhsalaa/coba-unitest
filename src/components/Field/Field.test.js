import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/react';

import { Field } from './index';
import { findByTestAttr } from '../../helpers/test';

describe('Field component', () => {
  describe('Component Renders', () => {
    it('Renders correctly and match snapshot', () => {
      const wrapper = mount(<Field />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Renders error message if provided', () => {
      const wrapper = shallow(<Field error="salah bos" />);
      const error = findByTestAttr(wrapper, 'qa-input-error');

      expect(error.text()).toBe('salah bos');
    });

    it('Renders label if provided', () => {
      const wrapper = shallow(<Field label="Masukkan" />);
      const label = findByTestAttr(wrapper, 'qa-input-label');

      expect(label.text()).toBe('Masukkan');
    });
  });

  describe('Component Functionality', () => {
    it('should have value typed', async () => {
      const wrapper = render(<Field />);
      const input = wrapper.getByTestId('qa-input-field');

      await userEvent.type(input, 'hola como estas');

      expect(input.value).toBe('hola como estas');
    });

    it('should not response value typed if readOnly', async () => {
      const wrapper = render(<Field readOnly />);
      const input = wrapper.getByTestId('qa-input-field');

      await userEvent.type(input, 'hola como estas');

      expect(input.value).toBe('');
    });

    it('should not response value typed if disabled', async () => {
      const wrapper = render(<Field disabled />);
      const input = wrapper.getByTestId('qa-input-field');

      await userEvent.type(input, 'hola como estas');

      expect(input.value).toBe('');
    });
  });

  describe('Component style', () => {
    it('should render grey background if readOnly', () => {
      const wrapper = mount(<Field readOnly />);
      const field = findByTestAttr(wrapper, 'qa-input-field');

      expect(field.props().className).toContain('bg-gray-100');
    });

    it('should render red border if error', () => {
      const wrapper = mount(<Field error={{ message: 'salah' }} />);
      const field = findByTestAttr(wrapper, 'qa-input-field');

      expect(field.props().className).toContain('border-red-700');
    });
  });
});
