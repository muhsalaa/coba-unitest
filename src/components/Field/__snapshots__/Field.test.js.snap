// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component Component Renders Renders correctly and match snapshot 1`] = `
<Field
  label={false}
  type="text"
>
  <div
    className="mb-4"
    data-testid="qa-input-wrapper"
  >
    <input
      className="h-12 border rounded w-full py-2 px-3 leading-tight focus:outline-none focus:border-blue-300"
      data-testid="qa-input-field"
      type="text"
    />
  </div>
</Field>
`;
